lib-src/libcs/run.3
.\" COPYRIGHT NOTICE
.\" Copyright (c) 1994 Carnegie Mellon University
.\" All Rights Reserved.
.\" 
.\" See <cmu_copyright.h> for use and distribution information.
.\" 
.\" 
.\" HISTORY
.\" $Log: run.3,v $
.\" Revision 4.1  1997/01/08 21:54:02  rvb
.\" r = 4.1; fix $ HEADERS
.\"
.\" Revision 1.1  1996/11/22 19:19:11  braam
.\" First Checkin (pre-release)
.\"
Revision 1.1  96/06/03  19:01:03  satya
Initial revision

.\" Revision 1.3  1995/10/19  20:47:38  moore
.\" 	Added some vfork handling
.\" 	[1995/10/19  20:46:54  moore]
.\"
.\" Revision 1.2  1995/10/18  14:59:00  moore
.\" 	Created libcs man(3) pages from mach entries
.\" 	[1995/10/18  14:55:43  moore]
.\" 
.\" $EndLog$
.\" Copyright (c) 1990 Carnegie Mellon University
.\" All Rights Reserved.
.\" 
.\" Permission to use, copy, modify and distribute this software and its
.\" documentation is hereby granted, provided that both the copyright
.\" notice and this permission notice appear in all copies of the
.\" software, derivative works or modified versions, and any portions
.\" thereof, and that both notices appear in supporting documentation.
.\"
.\" THE SOFTWARE IS PROVIDED "AS IS" AND CARNEGIE MELLON UNIVERSITY
.\" DISCLAIMS ALL WARRANTIES WITH REGARD TO THIS SOFTWARE, INCLUDING ALL
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS.  IN NO EVENT
.\" SHALL CARNEGIE MELLON UNIVERSITY BE LIABLE FOR ANY SPECIAL, DIRECT,
.\" INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER
.\" RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF
.\" CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF OR IN
.\" CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
.\"
.\" Users of this software agree to return to Carnegie Mellon any
.\" improvements or extensions that they make and grant Carnegie the
.\" rights to redistribute these changes.
.\"
.\" Export of this software is permitted only after complying with the
.\" regulations of the U.S. Deptartment of Commerce relating to the
.\" Export of Technical Data.
.\"""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""
.\" HISTORY
.\" $Log: run.3,v $
.\" Revision 4.1  1997/01/08 21:54:02  rvb
.\" r = 4.1; fix $ HEADERS
.\"
.\" Revision 1.1  1996/11/22 19:19:11  braam
.\" First Checkin (pre-release)
.\"
Revision 1.1  96/06/03  19:01:03  satya
Initial revision

.\" Revision 1.3  1995/10/19  20:47:38  moore
.\" 	Added some vfork handling
.\" 	[1995/10/19  20:46:54  moore]
.\"
.\" Revision 1.2.1.2  1995/10/19  20:46:54  moore
.\" 	Added some vfork handling
.\"
.\" Revision 1.2  1995/10/18  14:59:00  moore
.\" 	Created libcs man(3) pages from mach entries
.\" 	[1995/10/18  14:55:43  moore]
.\"
.\" Revision 1.1.1.2  1995/10/18  14:55:43  moore
.\" 	Created libcs man(3) pages from mach entries
.\"
.\" Revision 1.3  90/12/12  15:56:41  mja
.\" 	Add copyright/disclaimer for distribution.
.\" 
.\" Revision 1.2  90/02/01  16:14:52  mja
.\" 	Update to describe runc() and friends.
.\" 	[90/02/01  15:51:37  mja]
.\" 
.\" 13-Nov-86  Andi Swimmer (andi) at Carnegie-Mellon University
.\"	 Revised for 4.3.
.\" 
.\" 15-Jul-82  Neal Friedman (naf) at Carnegie-Mellon University
.\" 	Mentioned -1 return for vfork() failure.
.\" 
.\" 28-Jan-80  Steven Shafer (sas) at Carnegie-Mellon University
.\" 	Added setuid and setgid feature, so that run may be used
.\" 	by privileged programs (e.g. post) to execute programs
.\"	 which allow users to fork shells (e.g. text editors).
.\" 
.\" 21-Jan-80  Steven Shafer (sas) at Carnegie-Mellon University
.\" 	Changed  fork () to vfork ().  This wins speed if run (etc.) is
.\" 	 called from inside a very large program.
.\" 
.\" 05-Dec-79  Steven Shafer (sas) at Carnegie-Mellon University
.\" 	Created.
.\" 
.TH RUN 3 2/1/90
.CM 4
.SH NAME
run, runv, runp, runvp, runc, runcv, runcp, runcvp \- execute process and wait for it
.SH SYNOPSIS
.B 
int run (file,arg0,arg1,arg2,...,argn,0);
.br
.B 
int runv (file,arglist);
.br
.B 
int runp (file,arg0,arg1,arg2,...,argn,0);
.br
.B 
int runvp (file,arglist);
.PP
.B 
char *file,*arg1,*arg2,...,*argn,**arglist;
.sp
.B 
int runc (func,file,arg0,arg1,arg2,...,argn,0);
.br
.B 
int runcv (func,file,arglist);
.br
.B 
int runcp (func,file,arg0,arg1,arg2,...,argn,0);
.br
.B 
int runcvp (func,file,arglist);
.sp
.B 
char *file,*arg1,*arg2,...,*argn,**arglist;
.br
int (*func)();
.SH DESCRIPTION
.I 
Run
and
.I 
runv
have argument lists identical to the corresponding
functions,
.IR execl (2)
and
.IR execp (2).
The run routines perform a
.IR vfork (2),
or
.IR fork(2),
depending upon the machine type and operating system
then:
.TP 
In the new process:
.IR setgid (2)
and
.IR setuid (2)
are used to ensure that privileges unique to the parent do not
propagate to the child.
An
.I 
execl
or
.I 
execv
is then performed with the specified arguments.
The process returns
with a \-1 code if the
.I 
exec
was not successful.
.TP 
In the parent process:
the signals
.I 
SIGQUIT
(see
.IR signal (2))
and
.I 
SIGINT
are disabled, the process
.I 
waits
(see
.IR wait (2))
until the newly forked process exits, the signals
are restored to their original status, and the
return status of the process is analyzed.
.i0
.DT
.PP
.I 
Run
and
.I 
runv
return \-1 if 
.IR vfork/fork (2) 
or 
.IR exec(2) fail or the child was terminated
by a
.I 
signal;
the
.I 
exit
code of the process otherwise.

.I 
Runp
and
.I 
runvp
are identical to
.I 
run
and
.I 
runv,
but perform path searching for the process by using
.I 
execlp
and
.I 
execvp.
These routines use the PATH environment parameter
as a list of directory names separated by colons;
the executable file is sought in each directory
until it is found or all directories have been
searched.
If the file is not found, \-1 is returned.

The proper way to execute system programs is via
.I 
runp
or
.I 
runvp
for most purposes; for example, if you want to
move file "a" to "b", the best way to do this via
system programs is this:

.nf
.B	runp ("mv","mv","a","b",0);
.fi

Note that no directory name is needed along with the
name of the file (e.g.
.I /bin/mv 
is not necessary),
and that the program name should be both
.I 
file
and
.I 
arg0.
This call is similar to:
.nf
	system ("mv a b");
.fi
but is much faster to execute.

.I Runc,
.I runcp,
.I runcv,
and
.I runcvp
function analagously to their similarly named counterparts described
above.
However, they also invoke the specified function
.I func
in the context of the child process
in order to first perform any application
specific initialization
that may be needed before attempting to execute the program.
If
.I func
is null,
these routines are identical to their simpler counterparts.

The use of
.I 
setgid
and
.I 
setuid
means that, if the parent process gained privileges through the
use of special file mode bits (see 
.IR chmod (2)),
the child process
will not inherit these privileges.
This makes
.I 
run
"safe" for system programs which require special privileges, and
usually has no effect on user programs.
.SH ENVIRONMENT
The
.B 
PATH
environment parameter is used to find executable files in
.I 
runp
and
.I 
runvp.
If this parameter is not present, then the default value
.I 
:/usr/cs/bin:/bin:/usr/bin
is used.
.SH "LIBRARY NOTES"
You must link against /usr/local/lib/libcs.a to define these routines.
.SH "SEE ALSO"
.IR exec (2), 
.IR fork (2), 
.IR signal (2), 
.IR system (3), 
.IR searchp (3)
.SH DIAGNOSTICS
These routines return \-1 if any error occurs in executing
the desired program.
If the program is executed successfully,
convention dictates that it should return 0 on normal
completion and non-zero (1, 2, etc.) if any error is
encountered.
.SH BUGS
The searching rule used by
.I 
execlp
and
.I 
execvp
is not the same as the rule used by
.IR searchp .
See the comments in
.IR searchp (3)
for more detailed information.


NEWS
Incompatible RVM update (Coda client)
-------------------------------------

2008/02/06 RVM version 39
    Move store identifier from per-volume to global rvg struct.

2005/06/07 RVM version 38
    Added realm::realmid (as we can't use &realm_object on 64-bit systems)

2005/06/16 RVM version 37
    Removed fsdb::device and fsobj::inode


lib-src/libcs/run.c
#ifndef _BLURB_
#define _BLURB_
/*
 * This code was originally part of the CMU SCS library "libcs".
 * A branch of that code was taken in May 1996, to produce this
 * standalone version of the library for use in Coda and Odyssey
 * distribution.  The copyright and terms of distribution are
 * unchanged, as reproduced below.
 *
 * Copyright (c) 1990-96 Carnegie Mellon University
 * All Rights Reserved.
 * 
 * Permission to use, copy, modify and distribute this software and its
 * documentation is hereby granted, provided that both the copyright
 * notice and this permission notice appear in all copies of the
 * software, derivative works or modified versions, and any portions
 * thereof, and that both notices appear in supporting documentation.
 *
 * THE SOFTWARE IS PROVIDED "AS IS" AND CARNEGIE MELLON UNIVERSITY
 * DISCLAIMS ALL WARRANTIES WITH REGARD TO THIS SOFTWARE, INCLUDING ALL
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS.  IN NO EVENT
 * SHALL CARNEGIE MELLON UNIVERSITY BE LIABLE FOR ANY SPECIAL, DIRECT,
 * INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER
 * RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF
 * CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF OR IN
 * CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 *
 * Users of this software agree to return to Carnegie Mellon any
 * improvements or extensions that they make and grant Carnegie the
 * rights to redistribute these changes.
 *
 * Export of this software is permitted only after complying with the
 * regulations of the U.S. Deptartment of Commerce relating to the
 * Export of Technical Data.
 */

static char *rcsid = "$Header: /usr/rvb/XX/src/lib-src/libcs/RCS/run.c,v 4.1 1997/01/08 21:54:02 rvb Exp $";
#endif /*_BLURB_*/

/*  run[c], run[c]v, run[c]p, run[c]vp -- exec process and wait for it to exit
 *
 *  Usage:
 *	i = run (file, arg1, arg2, ..., argn, 0);
 *	i = runv (file, arglist);
 *	i = runp (file, arg1, arg2, ..., argn, 0);
 *	i = runvp (file, arglist);
 *	i = runc (func, file, arg1, arg2, ..., argn, 0);
 *	i = runcv (func, file, arglist);
 *	i = runcp (func, file, arg1, arg2, ..., argn, 0);
 *	i = runcvp (func, file, arglist);
 *
 *  Run, runv, runp, runvp and runc, runcv, runcp, runcvp have argument lists
 *  exactly like the corresponding routines, execl, execv, execlp, execvp.  The
 *  run routines perform a fork, then:
 *  IN THE NEW PROCESS, an execl[p] or execv[p] is performed with the specified
 *  arguments (after first invoking the supplied function in the runc* cases).
 *  The process returns with a -1 code if the exec was not successful.
 *  IN THE PARENT PROCESS, the signals SIGQUIT and SIGINT are disabled,
 *  the process waits until the newly forked process exits, the
 *  signals are restored to their original status, and the return
 *  status of the process is analyzed.
 *  All run routines return:  -1 if the exec failed or if the child was
 *  terminated abnormally; otherwise, the exit code of the child is
 *  returned.
 *
 */

#include <sys/types.h>
#include <sys/wait.h>
#include <sys/time.h>
#include <sys/resource.h>
#include <stdio.h>
#include <signal.h>
#include <unistd.h>

#if	__STDC__
#include <stdarg.h>
#define	VA_START(a, b) va_start(a, b)
#else
#include <varargs.h>
#define	VA_START(a, b) va_start(a)
#endif

#include "libcs.h"

#define RUN_MAXARGS 1024

#if defined(__linux__) || defined(__BSD44__) || defined(__WIN32__)
#define	sigif	sigaction
#define	sif_handler	sa_handler
#define	sif_mask	sa_mask
#define	sif_flags	sa_flags
#else
#define	sigif	sigvec
#define	sif_handler	sv_handler
#define	sif_mask	sv_mask
#define	sif_flags	sv_flags
#endif

static
int dorun (func,name,argv,usepath)
int (*func)();
const char *name;
char **argv;
int usepath;
{
	int wpid;
	register int pid;
#ifdef	__linux__
	struct sigaction ignoresig,intsig,quitsig;
#else
	struct sigif ignoresig,intsig,quitsig;
#endif
#if	defined(__BSD44__) || defined(__WIN32__)
	int status;
#else
	union wait status;
#endif
	int execvp(), execv();

#if defined(__sgi) || defined(_AIX)
#define vfork() fork()
#endif
	if ((pid = vfork()) == -1)
		return(-1);	/* no more process's, so exit with error */

	if (pid == 0) {			/* child process */
		if (func)
		    (*func)();
		(void) setgid (getgid());
		(void) setuid (getuid());
		(*(usepath ? execvp : execv)) (name,argv);
		fprintf (stderr,"run: can't exec %s\n",name);
		_exit (0377);
	}
#ifdef	__linux__
	ignoresig.sa_handler = SIG_IGN;	/* ignore INT and QUIT signals */
#else
	ignoresig.sif_handler = SIG_IGN;	/* ignore INT and QUIT signals */
#endif
#if	defined(__BSD44__) || defined(__WIN32__)
	sigemptyset(&ignoresig.sif_mask);
	ignoresig.sif_flags = 0;
#elif	defined(__linux__)
	ignoresig.sa_mask = 0;
	ignoresig.sa_flags = 0;
#else
	ignoresig.sif_mask = 0;
	ignoresig.sif_flags = 0;
#endif
	(void) sigif (SIGINT,&ignoresig,&intsig);
	(void) sigif (SIGQUIT,&ignoresig,&quitsig);

	do {
#if	defined(__BSD44__) || defined(__WIN32__)
		wpid = waitpid (-1, &status, WUNTRACED);
#else
		wpid = wait3 (&status, WUNTRACED, (struct rusage *)0);
#endif
		if (WIFSTOPPED (status)) {
		    (void) kill (0,SIGTSTP);
		    wpid = 0;
		}
	} while (wpid != pid && wpid != -1);

	/* restore signals */
	(void) sigif (SIGINT,&intsig,(struct sigif *)0);
	(void) sigif (SIGQUIT,&quitsig,(struct sigif *)0);
#if	defined(__BSD44__) || defined(__WIN32__)
	if (WIFSIGNALED (status) ||
	    (WIFEXITED(status) && WEXITSTATUS(status) == 0xff))
		return (-1);

	return (WEXITSTATUS(status));
#else
	if (WIFSIGNALED (status) || status.w_retcode == 0377)
		return (-1);

	return (status.w_retcode);
#endif
}

int runv (name,argv)
const char *name;
char **argv;
{
	return (dorun ((int (*)())0, name, argv, 0));
}

int runvp (name,argv)
const char *name;
char **argv;
{
	return (dorun ((int (*)())0, name, argv, 1));
}

int 
#if	__STDC__
runp (const char *name, ...)
#else
runp (name,va_alist)
const char *name;
va_dcl
#endif
{
	va_list ap;
	int val;
	int i;
  	char *args[RUN_MAXARGS];

	VA_START(ap, name);
	for (i=0; i<RUN_MAXARGS; i++)
	  {
	    args[i] = va_arg(ap,char *);
	    if (args[i] == (char *) 0)
	      break;
	  }

	if (i == RUN_MAXARGS)
	    return -1;

	val = runvp (name, args);
	va_end(ap);

	return (val);
}

int runcv (func,name,argv)
int (*func)();
const char *name;
char **argv;
{
	return (dorun (func, name, argv, 0));
}

int
#if	__STDC__
runc (int (*func)(), const char *name, ...)
#else
runc (func,name,va_alist)
int (*func)();
const char *name;
va_dcl
#endif
{
	va_list ap;
	int val;
	int i;
  	char *args[RUN_MAXARGS];

	VA_START(ap, name);
	for (i=0; i<RUN_MAXARGS; i++)
	  {
	    args[i] = va_arg(ap,char *);
	    if (args[i] == (char *) 0)
	      break;
	  }

	if (i == RUN_MAXARGS)
	    return -1;

	va_end(ap);
	val = runcv (func, name, args);
	return(val);
}

int runcvp (func,name,argv)
int (*func)();
const char *name;
char **argv;
{
	return (dorun (func, name, argv, 1));
}

int
#if	__STDC__
runcp (int (*func)(), const char *name, ...)
#else
runcp (func,name,va_alist)
int (*func)();
const char *name;
va_dcl
#endif
{
	va_list ap;
	int val;
	int i;
  	char *args[RUN_MAXARGS];

	VA_START(ap, name);
	for (i=0; i<RUN_MAXARGS; i++)
	  {
	    args[i] = va_arg(ap,char *);
	    if (args[i] == (char *) 0)
	      break;
	  }

	if (i == RUN_MAXARGS)
	    return -1;
	val = runcvp (func, name, args);
	va_end(ap);
	return (val);
}

int
#if	__STDC__
run (const char *name, ...)
#else
run (name,va_alist)
const char *name;
va_dcl
#endif
{
	va_list ap;
	int val;
	int i;
  	char *args[RUN_MAXARGS];

	VA_START(ap, name);
	for (i=0; i<RUN_MAXARGS; i++)
	  {
	    args[i] = va_arg(ap,char *);
	    if (args[i] == (char *) 0)
	      break;
	  }

	if (i == RUN_MAXARGS)
	    return -1;

	val = runv (name, args);
	va_end(ap);
	return(val);
}
